articles/integrating-b2c-mpesa-api-using-node-js/index.md

```diff
@@ -0,0 +1,299 @@
+
+    // set the status code.
+    res.status(200);
+
+    // send the access token to client
+    return res.send({
+        accessToken
+    });
+};
+```
+
+From above:
+
+- Destructuring the `consumerkey` and `consumersecret` from the `headers`.
+- Setting the URL to send the request to.
+- Generating a `base64` encoded string from a buffer.
+- Composing the authentication string by appending `Basic` before the encoded string.
+- Sending the request to Mpesa API from a `try/catch` block. In case of an error, the error is sent to the client. Else if there is no error, the access token is sent to the client.
+
+To test it:
+
+- Start the development server by running the following command from the terminal of your text editor:
+
+```bash
+npm run dev
+```
+
+- In your browser, open `http://localhost:4000/api-docs`.
+- Click the `/access-token` section under `Mpesa B2C` tag.
+- Click the `Try it out` button.
+- To get your `consumer key` and `consumer secret`, visit your [applications page](https://developer.safaricom.co.ke/user/me/apps), select your application, from the keys section copy and paste them respectively into the section's parameters.
+- Click the `Execute` button.
+- In case of an error, revisit the steps. Else your server response should resemble the following:
+
+![access_token_response_screenshot](/engineering-education/integrating-b2c-mpesa-api-using-node-js/access-token-response-screenshot.png)
+
+_A screenshot of getting access token server response_
+
+### B2c API
+
+With the access token functionality implemented, we are set to implement the B2C API.
+To do that, we need to modify the `b2c()` method as follows:
+
+```js
+//b2c.
+async b2c(req,res,next){
+// destructure from headers
+let {accesstoken,initiatorname,securitycredential,commandid,amount,partya,partyb,remarks}= req.headers;
+// url to send request to.
+let url = "https://sandbox.safaricom.co.ke/mpesa/b2c/v1/paymentrequest";
+// url to expose our local server
+let ngrok_url = process.env.NGROK_URL;
+// authentication string
+let auth = `Bearer ${accesstoken}`;
+let response;
+
+try{
+
+    // send the request
+    response = await axios.default.post(url,{
+    "InitiatorName": initiatorname,
+    "SecurityCredential":securitycredential,
+    "CommandID": commandid,
+    "Amount": amount,
+    "PartyA": partya,
+    "PartyB": partyb,
+    "Remarks": remarks,
+    "QueueTimeOutURL": `${ngrok_url}/timeout`,
+    "ResultURL": `${ngrok_url}/cb`,
+    "Occasion": remarks
+    },{
+        headers:{
+            "Authorization":auth
+        }
+    })
+
+}catch(error) {
+
+    // in case of an error, get the code and the message.
+    let err_code = error.response.status;
+    let err_msg = error.response.data.errorMessage;
+
+    // send to the client
+    return res.status(err_code).send({
+        message:err_msg
+    });
+}
+
+// set the status code
+res.status(200);
+
+// send to the client
+return res.send({
+    result:response.data
+});
+};
+```
+
+From above:
```

Comment:
Could you explain more from the above code

Reply:
done